### PR TITLE
Sc 40993 requirements update

### PR DESCRIPTION
**Pull request by rashidnhm**
**Title:** Migrate dependencies to python-poetry

**Summary:**
This PR brings a big change to the QML workflow.

**This issue to address**
The current state of the requirements are very difficult to work with. It is brittle, and the usage of the many requirements file is un-intuitive to grasp.

Furthermore, due to pip's limitation on how it tracks dependencies, it is difficult to determine which packages depend on which other packages. 

**Proposed solution in this PR**
This PR removes all the previous requirements files and introduces a new file in the root of the project called `pyproject.toml` (and the accompanying poetry.lock file as well).

The poetry.lock file tracks all the dependencies similar to the previous requirements file, but the toml file now only tracks the top level dependency.

**Relevant references:**
[sc-40993](https://app.shortcut.com/xanaduai/story/40993/update-qml-dependencies-to-work-with-python-3-10) (This is the first phase of the QML dependency update work, but this PR itself does not update any version of any dependency).

**Possible Drawbacks:**
As this is a major change, it will something new contributors will need to get used to. We will also need to iron out any issues we hit with poetry that did not arise during the testing of this PR.

**Related GitHub Issues:**
None.

### CODE REVIEW

File: build_requirements/master/poetry.lock [hunk 1]
```diff
@@ -0,0 +1,5274 @@
...
+    {file = "pyscf-2.2.1-cp38-cp38-macosx_10_15_x86_64.whl", hash = "sha256:ff96e85977d7bf5017b19c6de37f87129c888e1c64d895be613cc955f7e748ff"},
+    {file = "pyscf-2.2.1-cp38-cp38-manylinux_2_17_aarch64.manylinux2014_aarch64.whl", hash = "sha256:985a94bd34e15ede01a90d5d44d2e0f3c4946603c2ad887559a434c0385a0633"},
+    {file = "pyscf-2.2.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:69f3dc501a71d2b4c0e98ca028b42684e8e16b548ec1e6fe4978fde8c0786f7d"},
+    {file = "pyscf-2.2.1-cp39-cp39-macosx_11_0_x86_64.whl", hash = "sha256:365123d0314257e720c038ec51bd09a87fcfdb76ab5a7612543cb851292a80db"},
+    {file = "pyscf-2.2.1-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl", hash = "sha256:6e7bcdda255b19cf5f95863d1537a98fb6c561927f582dad9aac0203f29f8d5d"},
+    {file = "pyscf-2.2.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:ee89c5a9e7e10482da10d15f54801ef4312971f5c9e87361dd2e68eb036714f0"},
+    {file = "pyscf-2.2.1.tar.gz", hash = "sha256:4ff6851351caadc5dfa543b6b2c5fbd926ded87e3cc39faa0054e1e5090ed69a"},
+]
+
+[package.dependencies]
```
Comment (timmysilv) [inline build_requirements/master/poetry.lock:4449]: this failed to install on my computer (MacBook, ARM), but I don't think it was working before either

File: .github/workflows/build-branch.yml [hunk 2]
```diff
@@ -209,15 +209,45 @@ jobs:
...
+      - name: Set up Python Version
+        id: setup_python
+        uses: actions/setup-python@v4
+        if: needs.compute-build-strategy-matrix.outputs.build-environment == 'master'
+        with:
+          python-version: 3.9
+
+      - name: Install Poetry
+        if: needs.compute-build-strategy-matrix.outputs.build-environment == 'master'
+        run: python3 -m pip install poetry==1.5.1
```
Comment (timmysilv) [inline .github/workflows/build-branch.yml:239]: is this the recommended way to install it in a CI action?

File: .github/workflows/build-branch.yml [hunk 3]
```diff
@@ -252,12 +282,38 @@ jobs:
...
         run: |
-          ${{ steps.venv.outputs.location }}/bin/python3 -m pip install -r requirements.txt
-          ${{ steps.venv.outputs.location }}/bin/python3 -m pip install --no-deps -r requirements_no_deps.txt
-          ${{ steps.venv.outputs.location }}/bin/python3 -m pip install --index-url https://test.pypi.org/simple/ pennylane-lightning --pre --no-deps --upgrade
-          ${{ steps.venv.outputs.location }}/bin/python3 -m pip install --upgrade pip 'setuptools<65' cmake
+          source ${{ steps.venv.outputs.location }}/bin/activate
+          ${{ steps.venv.outputs.location }}/bin/python3 -m poetry -C build_requirements/${{ needs.compute-build-strategy-matrix.outputs.build-environment }} install
+          
+          # Adding this install outside of poetry as the dependencies conflict with existing other packages.
+          # TODO: This needs to be revisited.
```
Comment (timmysilv) [inline .github/workflows/build-branch.yml:309]: is this happening in this PR/being tracked somewhere?

File: __deprecated_requirements-norun.txt
Comment (timmysilv) [inline __deprecated_requirements-norun.txt:1]: do we need to keep this and other deprecated files around? the information still exists, plus there's git history. I'm ok to just delete them

File: build_requirements/dev/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+sphinx = "3.5.3"
+sphinx-sitemap = "*"
+sphinx_gallery = "0.10.0"
+Jinja2 = "2.11.3"
+markupsafe = "2.0.1"
+numpy = "~1.21"
+pyyaml = "5.4.1"
+pennylane-sphinx-theme = "*"
+pypandoc = "1.5"
+docutils = ">=0.12,<0.17"  # Sphinx breaks if >= 0.17 is installed
```
Comment (timmysilv) [inline build_requirements/dev/pyproject.toml:34]: unrelated but surely we can unpin this by now

File: build_requirements/dev/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,65 @@
+[tool.poetry]
+name = "qml"
+version = "0.0.0-dev"
+description = "Introductions to key concepts in quantum machine learning, as well as tutorials and implementations from cutting-edge QML research."
+authors = ["PennyLaneAI <info@pennylane.ai>"]
+license = "Apache-2.0"
+readme = "README.md"
```
Comment (timmysilv) [inline build_requirements/dev/pyproject.toml:7]: unless it was skipped on purpose
```suggestion
readme = "README.md"
homepage = "https://pennylane.ai/qml/"
```

File: build_requirements/master/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,66 @@
+[tool.poetry]
+name = "qml"
+version = "0.0.0"
```
Comment (timmysilv) [inline build_requirements/master/pyproject.toml:3]: should this be maintained/updated in some way by qml owners? is it published anywhere?

File: .github/workflows/build-branch.yml (hunk 2 above)
Comment (rashidnhm) [reply .github/workflows/build-branch.yml:239]: The official documentation mentions various ways to install poetry in a CI environment, and yes, this being one of them:

https://python-poetry.org/docs/#ci-recommendations

image

Comment (rashidnhm) [reply .github/workflows/build-branch.yml:239]: That being said, I do install poetry alongside the other deps in the same env, maybe poetry should be separate :thinking: ... will update that

File: .github/workflows/build-branch.yml (hunk 3 above)
Comment (rashidnhm) [reply .github/workflows/build-branch.yml:309]: Kind of .. the issue is scipy being pinned to <=0.10 on pennylane. mitiq depends on [scipy being 0.10.x](https://github.com/unitaryfund/mitiq/blob/v0.25.0/requirements.txt#L2)

Once scipy is unpinned on pennylane, this workaround can be removed and mitiq can be part of the poetry installation

File: build_requirements/master/pyproject.toml (hunk 6 above)
Comment (rashidnhm) [reply build_requirements/master/pyproject.toml:3]: No `0.0.0` is on purpose. QML is not a package we publish, it really just is a website that has a rolling release.

File: build_requirements/dev/pyproject.toml (hunk 5 above)
Comment (rashidnhm) [reply build_requirements/dev/pyproject.toml:7]: Ah whoops, I'll add that back

File: build_requirements/dev/pyproject.toml (hunk 4 above)
Comment (rashidnhm) [reply build_requirements/dev/pyproject.toml:34]: We can remove docutils from the pyproject.toml altogether once we upgrade the version of sphinx we use. However, the version of sphinx we use has a bug where is lists any version of docutils >= 0.12 is comptabile with it ... except it isn't and sphinx rendering breaks if docutils 0.17 or greater is installed. This bug was fixed in a newer version of sphinx.

File: build_requirements/master/poetry.lock (hunk 1 above)
Comment (rashidnhm) [reply build_requirements/master/poetry.lock:4449]: What version of Python are you running?

Comment (timmysilv) [reply build_requirements/master/poetry.lock:4449]: 3.9.13. does this need 3.10 perhaps?

File: README.md [hunk 7]
```diff
@@ -187,6 +187,64 @@ there are a couple of guidelines to keep in mind.
...
+### Adding / Editing dependencies
+
+All dependencies need to be added to the pyproject.toml. It is recommended that unless necessary, 
+all dependencies be pinned to as tight of a version as possible.
+
+Add the new dependency in the `[tool.poetry.group.executable-dependencies.dependencies]` section of the toml file.
+
+Once pyproject.toml files have been updated, the poetry.lock file needs to be refreshed:
+```bash
+poetry lock
```
Comment (KetpuntoG) [inline README.md:230]: Thanks for open this! Where do we run this? Or is it something that checks a github action?

Comment (rashidnhm) [reply README.md:230]: You have to run it locally on your machine. Links to install `poetry` has been added to the README as well: https://github.com/PennyLaneAI/qml/blob/sc-40993-requirements-update/README.md?plain=1#L194-L195

Comment (rashidnhm) [reply README.md:230]: Here is the link to installing poetry: https://python-poetry.org/docs/#installation

Anyone working with the QML repo locally will now need this tool installed.

File: .github/workflows/build-branch.yml [hunk 8]
```diff
@@ -238,7 +267,7 @@ jobs:
         uses: actions/cache@v3
         with:
           path: venv
-          key: pip-${{ steps.setup_python.outputs.python-version }}-${{ hashFiles('requirements.txt') }}-${{ hashFiles('requirements_no_deps.txt') }}
+          key: pip-${{ steps.setup_python.outputs.python-version }}-${{ hashFiles('build_requirements/**/pyproject.toml') }}
```
Comment (timmysilv) [inline .github/workflows/build-branch.yml:270]:
```suggestion
          key: pip-${{ steps.setup_python.outputs.python-version }}-${{ hashFiles('pyproject.toml') }}
```

File: README.md [hunk 9]
```diff
@@ -58,9 +58,9 @@ quantum computing paper/result.
 ### Adding demos
 
 - Demos are written in the form of an executable Python script.
-  Any package listed in `requirements.txt` and `requirements_no_deps.txt` you can assume is
-  available to be imported.
-  Matplotlib plots will be automatically rendered and displayed on the QML website.
+  - Packages listed in `build_requirements/master/pyproject.toml` (or dev) will be available for import during execution.
```
Comment (timmysilv) [inline README.md:61]:
```suggestion
  - Packages listed in `pyproject.toml` will be available for import during execution.
```

File: README.md [hunk 10]
```diff
@@ -187,6 +187,64 @@ there are a couple of guidelines to keep in mind.
...
+
+To install all the dependencies locally, [poetry](https://python-poetry.org/) needs to be installed. Please follow the
+[official installation documentation](https://python-poetry.org/docs/#installation).
+
+### Installing dependencies
+
+Once poetry has been installed, the dependencies can be installed as follows:
+```bash
+make environment
+```
```
Comment (timmysilv) [inline README.md:202]: I would add a line below this mentioning the availability of `POETRYOPTS` to pass options along to `poetry install`

File: README.md [hunk 11]
```diff
@@ -216,18 +274,20 @@ To install dependencies on an M1 Mac and build the QML website, the following in
...
   ```
 
-- Install each package in `requirements-norun.txt` by running
+- Follow the steps from `Dependency Management` to setup poetry.
+
+- Install the base packages by running
 
   ```bash
-  pip3 install -r requirements-norun.txt
+  poetry -C build_requirements/master install --without executable-dependencies
```
Comment (timmysilv) [inline README.md:282]:
```suggestion
  make environment BASE_ONLY=true
```

File: README.md [hunk 12]
```diff
@@ -216,18 +274,20 @@ To install dependencies on an M1 Mac and build the QML website, the following in
...
   ```
 
   Alternatively, you can do this in a new virtual environment using
 
   ```bash
   python -m venv [venv_name]
   cd [venv_name] && source bin/activate
-  pip install -r requirements-norun.txt
+  poetry -C build_requirements/master install --without executable-dependencies
```
Comment (timmysilv) [inline README.md:290]:
```suggestion
  make environment BASE_ONLY=true
```

File: Makefile [hunk 13]
```diff
@@ -8,6 +8,12 @@ SOURCEDIR     = .
...
 DATADIR       = _data
 
+# Environment setup variables
+POETRY_BIN     = poetry
+POETRYOPTS     =
+UPGRADE_PL     = false
+BASE_ONLY      = false
+
 # Put it first so that "make" without argument is like "make help".
 help:
```
Comment (timmysilv) [inline Makefile:18]: can you add a one-liner in the help target that describes the new `environment` target?

File: Makefile [hunk 14]
```diff
@@ -8,16 +8,25 @@ SOURCEDIR     = .
...
 	@$(SPHINXBUILD) -M help "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O)
+	@echo
+	@echo "Additional Targets:"
+	@echo "  environment      Setup all packages and dependencies needed to build the docs"
```
Comment (timmysilv) [inline Makefile:22]: oh I didn't realize the `help` target was from Sphinx - your addition works for me!

File: README.md [hunk 15]
```diff
@@ -187,6 +187,69 @@ there are a couple of guidelines to keep in mind.
...
+```
+Note: This makefile target calls `poetry install` under the hood, you can pass any poetry arguments to this by passing
+the `POETRYOPTS` variable.
+```bash
+make environment POETRYOPTS='--sync --dry-run --verbose'
+```
+
+The `master` branch of QML uses the latest stable release of PennyLane, whereas the `dev` branch uses the most 
+up-to-date version from the GitHub repository. If your demo relies on that, install the `dev` dependencies instead
+by upgrading all PennyLane and Plugins to the latest commit from GitHub.
```
Comment (timmysilv) [inline README.md:211]: I can't propose changes but can you make this say "by upgrading PennyLane and its various plugins"?

File: README.md [hunk 16]
```diff
@@ -187,6 +187,69 @@ there are a couple of guidelines to keep in mind.
...
+#### Adding / Editing PennyLane (or plugin) versions
+This process is slightly different from other packages. It is due to the fact that the `master` builds use the stable
+releases of PennyLane as stated in the pyproject.toml file. However, for dev builds, we use the latest commit from 
+GitHub.
+
+##### Adding a new PennyLane package (plugin)
+- Add the package to `pyproject.toml` file with the other pennylane packages and pin it to the latest stable release.
+- Add the GitHub installation link to the Makefile, so it is upgraded for dev builds with the other PennyLane packages.
+    - This should be under the format `$$PYTHON_VENV_PATH/bin/python -m pip install --upgrade git+https://github.com/PennyLaneAI/<repo>.git#egg=<repo>;\`
+- Refresh the poetry lock file `poetry lock`
```
Comment (timmysilv) [inline README.md:251]: again, can't propose changes for some reason, but "Refresh the poetry lock file by running `poetry lock`"

**Review by timmysilv — APPROVED** [review]
just 2 little doc changes, but otherwise lgtm!! 🎉 🎉  thanks for doing this

**Review by smokingroosters — APPROVED** [review]

**Pull request merged**